### PR TITLE
workaround pypi empty responses by disabling cache control

### PR DESCRIPTION
**Pull request by kapilt**
# Pull Request Check List

Resolves: #4821 
Alternative to: #4822 and #4717 

This is a work-around fix for https://github.com/pypa/warehouse/issues/10387 . Thanks @j-martin for also looking at this.

This disables cache control for package metadata due to issues observed that pypi can return empty responses. The package artifact is still maintained. The diff is a little larger due to need to keeping type checker happy and removing tests assuming cache control usage.

- [ x ] Added **tests** for changed code.
- [ x ] Updated **documentation** for changed code.

### CODE REVIEW

File: src/poetry/repositories/legacy_repository.py [hunk 1]
```diff
@@ -62,7 +62,9 @@ class Page:
         ".tar",
     ]
 
-    def __init__(self, url: str, content: str, headers: Dict[str, Any]) -> None:
+    def __init__(
+        self, url: str, content: bytes, headers: requests.structures.CaseInsensitiveDict
```
Comment (abn) [inline src/poetry/repositories/legacy_repository.py:66]: I would prefer that we stuck to `Dict` or `Mappting`. We do want to drop requests at some point.

File: src/poetry/repositories/pypi_repository.py [hunk 2]
```diff
@@ -72,18 +71,11 @@ def __init__(
         )
 
         self._cache_control_cache = FileCache(str(release_cache_dir / "_http"))
-        self._session = CacheControl(
```
Comment (abn) [inline src/poetry/repositories/pypi_repository.py:75]: I am not a 100% convinced this is the right solution for this. I'd much rather have e-tag used in the cache key.

Skipping cache for empty responses or invalid cache values, should be handled by gracefully handling failure - ie. clear the invalid cache and retry or fallback.

Comment (j-martin) [reply src/poetry/repositories/pypi_repository.py:75]: I guess that's the naive approach [here](https://github.com/python-poetry/poetry/pull/4717).

File: tests/fixtures/git/github.com/demo/demo/demo.egg-info/PKG-INFO [hunk 3]
```diff
@@ -6,7 +6,8 @@ Home-page: https://github.com/demo/demo
 Author: Sébastien Eustace
 Author-email: sebastien@eustace.io
 License: MIT
-Description: UNKNOWN
 Platform: UNKNOWN
-Provides-Extra: bar
 Provides-Extra: foo
+Provides-Extra: bar
+
```
Comment (neersighted) [inline tests/fixtures/git/github.com/demo/demo/demo.egg-info/PKG-INFO:12]: This is an invalid change that came along for the ride. Still need to find what test is causing this instability.

**Review by neersighted — CHANGES_REQUESTED** [review]

File: src/poetry/repositories/pypi_repository.py (hunk 2 above)
Comment (neersighted) [reply src/poetry/repositories/pypi_repository.py:75]: That approach is at the request level and not really at the cache level. I think an intermediate approach (e.g. layered on top of CacheControl and factoring in ETAGS) is the way to go. The caching is an important optimization, and adding a retry system at the high-level HTTP (e.g. `.get()`) API seems more a hack than a fix.

File: tests/fixtures/git/github.com/demo/demo/demo.egg-info/PKG-INFO (hunk 3 above)
Comment (kapilt) [reply tests/fixtures/git/github.com/demo/demo/demo.egg-info/PKG-INFO:12]: ack, yeah not sure where that line came from, will remove

File: src/poetry/repositories/legacy_repository.py (hunk 1 above)
Comment (kapilt) [reply src/poetry/repositories/legacy_repository.py]: switched to mapping, mypy was complaining about the dict.

**Comment by kapilt** [comment]
i haven't actually hit this in a while using latest poetry release, closing.

**Pull request closed without merging**